### ISSUE TITLE
KL-vanishing issue and another idea

### ISSUE THREAD

**Issue by Fanta007** [Mathux/ACTOR#5]
Hi Mathis
Thanks for sharing the implementation. I have two issues:
1. If I understand correctly, this work can be treated as a sequence VAE model used to learn distributions from observations. The KL-divergence loss term encourages the learned distribution to be as close as to a normal distribution. So have you ever encountered  the KL-vanishing (KL loss = 0) issue? If yes, how have you solved it ?

2. Instead of feeding the sampled code z merely to Transformer decoder block, how about to adding it to all tokens for the decoder input sequence (plus with the position embeddings)?

Thanks

**Comment by Mathux** [comment]
Hello Fanta007

> If I understand correctly, this work can be treated as a sequence VAE model used to learn distributions from observations. 

Yes, it is sequence level VAE; it encodes all the sequence into one representation which not depend one time.

> So have you ever encountered the KL-vanishing (KL loss = 0) issue? If yes, how have you solved it ?

At the beginning, I struggled a bit to train it, because I was looking for too high values for lambda_kl parameter, and so I had mode collapsing. When I tried very low value, like 1e-7, it begin to work well. 

> Instead of feeding the sampled code z merely to Transformer decoder block, how about to adding it to all tokens for the decoder input sequence (plus with the position embeddings)?

I think I already tried something similar during my prototypes, but it was at the beginning and it didn't work well. So maybe you should give a try!
